File: app/core/license.py
# -*- coding: utf-8 -*-

"""
Talvo Admin API - License Management
"""

import uuid
import hashlib
from datetime import datetime, timedelta


class LicenseManager:
    
    @staticmethod
    def generate_license_key(company_name: str, days_valid: int = 365) -> dict:
        unique_id = uuid.uuid4().hex.upper()
        expiry_date = (datetime.utcnow() + timedelta(days=days_valid)).isoformat()
        raw_key = f"{company_name}|{unique_id}|{expiry_date}"
        key_hash = hashlib.sha256(raw_key.encode()).hexdigest()[:16].upper()
        license_key = f"{unique_id[:8]}{key_hash}{unique_id[8:16]}"
        
        return {
            'license_key': license_key,
            'expiry_date': expiry_date,
            'company_name': company_name,
            'is_valid': True
        }
    
    @staticmethod
    def create_db_name(license_key: str) -> str:
        return f"tenant_{license_key[:8].lower()}"

File: src/utils/license.ts
import { randomUUID, createHash } from 'crypto';

export class LicenseManager {
  static generateLicenseKey(companyName: string, daysValid: number = 365) {
    const uniqueId = randomUUID().replace(/-/g, '').toUpperCase();
    const expiryDate = new Date();
    expiryDate.setDate(expiryDate.getDate() + daysValid);
    
    const rawKey = `${companyName}|${uniqueId}|${expiryDate.toISOString()}`;
    const keyHash = createHash('sha256')
      .update(rawKey)
      .digest('hex')
      .substring(0, 16)
      .toUpperCase();
    
    const licenseKey = `${uniqueId.substring(0,8)}${keyHash}${uniqueId.substring(8,16)}`;
    
    return {
      license_key: licenseKey,
      expiry_date: expiryDate.toISOString(),
      company_name: companyName,
      is_valid: true
    };
  }

  static createDbName(licenseKey: string): string {
    return `tenant_${licenseKey.substring(0, 8).toLowerCase()}`;
  }
}